### ISSUE TITLE
Exit after n minutes/n candles and roi = 0

### ISSUE THREAD

**Issue by gaugau3000** [freqtrade/freqtrade#2277]
Hi there,

1/ For testing purpose of a strategy i would like to do something like (test what happens if i close my trade after 30 minutes)

minimal_roi = {'30':-1}
stop_loss = -1

![image](https://user-images.githubusercontent.com/2643583/65308194-d1e59f80-db89-11e9-8ced-8aeb62f72bfc.png)

Freqtrade is not happy with that.

2/ this type of problem appear too when using 

minimal_roi = {'0': 0.02, '30': 0}
stoploss = -0.02

![image](https://user-images.githubusercontent.com/2643583/65308693-e5ddd100-db8a-11e9-9550-2be69f357bd9.png)

14% of the trades have 0% roi but it is not possible because when you close the trade 30 minutes with profit > 0 to must be take into consideration.

I think in live he don't have this type of problem ?

Have a nice day.

**Comment by xmatthias** [comment]
I think "sell-reason reporting" is the real problem?

your first configuration seems to work, but produces a huge loss after 30 minutes due to "roi". 
Ignore the sell-reason column for now and load the backtest-results file to pandas as shown in the documentation (i assume you're doing that here anyway, according to the screenshots...)

Assume you load the trades, then override sell_reason with `backtest_df.loc[backtest_df['profitperc'] < 0, 'sell_reason'] = "stoploss"`, and then do a `value_counts()` grouping by that column again (which will produce the same output than backtesting, but with fixed sell-reasons.

This should produce "some" positive results (maybe not - then your buy-signal is not suitable for 30min trades)...

**Comment by xmatthias** [comment]
Thinking about 2 again, i know the reason for this too:
backtesting always uses the "roi" value as sell-value. The reason for this is simple - assume your ROI setting is 2% - but "high" was 5% above buy price. 
in Dry/live runs, your bot would have sold at 2%, and not 5% - so using high is wrong, same with close (which could be 4% - but the bot sells as soon as 2% roi is hit).

Now in dry/live runs, you would not see 2% but 2.01% or so (depending on the sell ordertype, how fast the market moves etc) ... but simulating this reliably in backtesting is impossible, since we lack the information for it.

hope this helps explaining this.

**Comment by gaugau3000** [comment]
Hi,

the problematic i have is solve after n bar/n minutes no matter loss/profit. i would like to manage this with 

```
minimal_roi = {'30':-1}
stop_loss = -1
```

After doing that he stop the trade at 30 minutes but take the roi value : -1 as the profit (-99.9% per trade) and not the close-open value to calculate profit. 

![image](https://user-images.githubusercontent.com/2643583/65606138-9f351000-dfaa-11e9-9061-b1e72fa016d2.png)


2/ Here it is the same problem but with more realistic use case  i want to try closing the trade after 30 minutes :
54% make 2% profit before 30 minutes -> ok
31% touch the stop loss not depending time duration -> ok
14% of the trades are sold after 30 minutes if roi >= 0.
         -> ok if trade open since 30 minutes and current  loss is 1% -> at 40 minutes going up and roi 0 reach !
         -> nok if trade is winning at 30 minutes (current profit at bar close is 1.5%) but report as 0% in backtest.

```
 minimal_roi = {'0': 0.02, '30': 0}
 stoploss = -0.02
```

> ![image](https://user-images.githubusercontent.com/2643583/65308693-e5ddd100-db8a-11e9-9550-2be69f357bd9.png)

Many tks.

Have a nice day.

**Comment by xmatthias** [comment]
hi @gaugau3000 , 

sorry for the wait, you're right, this happens because we use the ROI value for sells (assuming it's within a candle).

i've tried to fix this in #2342. 
Would appreciate if you could have a look at the solution / give it a try with your strategy.
I've also created a dummy-testcase for this - but testing it against the strategy where it initially appeared can only be a benefit!

The solution is taking the maximum of the `calculated ROI price` and `low`, since we can't know the 100% correct value within the candle during backtesting.

There is no need to test against the candles High, since ROI is compared as `>` with the candle already, so having sells above the candle are not possible.

**Comment by gaugau3000** [comment]
@xmatthias many tks :-) your fix seem to work but let me the end of the week end to test (i will dry to dive of little into the code this time). I come back to you.

Have a nice WE.

**Comment by gaugau3000** [comment]
@xmatthias you can merge it is ok for me (don't have enough time do go into code but i make different test and it is ok)

Have a nice evening :-)

**Comment by gaugau3000** [comment]
Hi there,

I was thinking if it is possible to take the close of the candle instead of low when the roi time correspond to a candle close (only for negative roi and if stop loss not reach).

Example :

timeframe = 5m
30 is a multiple of 5 so take the close instead of low for computation (correspond to the close of the sixth candle).

> minimal_roi = {'30':-1}
> stop_loss = -1

Have a nice day.

**Comment by xmatthias** [comment]
is it possible - yes

is it realistic - i'm not certain (letme explain)
backtesting always trades "on the candle" - so you cannot have a trade that opens 10 seconds after the candle opened.

In real trading though, that's exactly what happens, you'll be 1s, 5s, maybe 10s late (depending on latency, amount of pairs, ...).
So instead of using close, should use open instead.

However, assuming the candle opens, and immediately drops to the candle's low - then using "low" in backtesting would be correct again - so the answer is not clear and i don't think it can be generalized well.

Would backtest results improve when using negative ROI and we'd switch to close or open? Yes sure
would it be better aligned? probably not...

**Comment by gaugau3000** [comment]
Hi

> is it possible - yes
> 
> is it realistic - i'm not certain (letme explain)
> backtesting always trades "on the candle" - so you cannot have a trade that opens 10 seconds after the candle opened.
> 
> In real trading though, that's exactly what happens, you'll be 1s, 5s, maybe 10s late (depending on latency, amount of pairs, ...).
> So instead of using close, should use open instead.

I agree.

> However, assuming the candle opens, and immediately drops to the candle's low - then using "low" in backtesting would be correct again - so the answer is not clear and i don't think it can be generalized well.

Why make this type of assumption ?

> Would backtest results improve when using negative ROI and we'd switch to close or open? Yes sure
> would it be better aligned? probably not...

It is ok for me to have a backtest conservative (and maybe it is not enough as spread is not taking into consideration). The only problem is i make stress test on strategies like : does this strategy makes money/have good winrate at the end of the first candle, the second... (it is a way to optimize buys)?

> ticker_interval = "5m"
> minimal_roi = {'5':-1}
> stop_loss = -1

and i am not able to to that for now because it takes the low

I know that it is a limit use case but takes the low is conservative but maybe not realistic (eg : end a trade after 1 candle in 4h timeframe takes the candle low).

Have a nice day.

**Comment by xmatthias** [comment]
> I know that it is a limit use case but takes the low is conservative but maybe not realistic (eg : end a trade after 1 candle in 4h timeframe takes the candle low).

well why assume you're able to sell at open rate? both are assumptions, which will at times happen, and at other times won't happen. 
using low is the more conservative approach.

We need to consider that this code does not only apply when using -1 as ROI.

Now let's assume the following case (these are 5m candles on BTG/ETH from a while ago):

Before this issue: (using ROI calculation alone (can lead to sells "outside" of the candle):
![2019-12-07-144937_521x455_scrot](https://user-images.githubusercontent.com/5024695/70375673-d146e480-1900-11ea-9348-df8846c37534.png)

Now:
![2019-12-07-145123_1040x481_scrot](https://user-images.githubusercontent.com/5024695/70375689-0d7a4500-1901-11ea-861b-7dc4b21fa28d.png)

Your desired future state:
![2019-12-07-145246_1104x559_scrot](https://user-images.githubusercontent.com/5024695/70375706-4c0fff80-1901-11ea-92a6-68c00512be7a.png)

leaving out the fact that it's a rather low-volume pair, so i should better stay away from this when trading real money:

I've been using ROI: of `"20": 0.01,` (the others are not relevant for this case).

In mathematical terms:
Since 120 is exactly a multiple of 5 - it should use the Open price according to you.

However:
the ROI setting is 0.01 (1%) ... 
currently, it sells at 0.013 (1.3%) - which is realistic (and assumes worst case) since price jumped - it would be impossible to sell at 1% since price has never been at that rate at this time.

by using open, it would show a sell at  0.015 (1.5%) - which might seem realistic if you're the first in the queue - but you are not (otherwise price would not jump - we did not act yet).

so it's not unrealistic to have price go down (possibility is just as high that price went up to high right away - but that would taint backesting towards assuming the best possible outcome).


To me, using low is the correct approach, unless we use some special code for "-1" roi setitngs.

**Comment by hroff-1902** [comment]
Assume we have buy order happened:
![0](https://user-images.githubusercontent.com/47309513/70853223-36c84180-1ebc-11ea-9834-49482fbd4e44.png)

In case of `minimal_roi = {"30": -1}`, it defines the "vertical barrier", which is at the beginning of the candle, the `open` price of that candle should be used:
![1](https://user-images.githubusercontent.com/47309513/70853246-71ca7500-1ebc-11ea-9c36-8023d7e0c54c.png)

If `minimal_roi = {"30": -0.9}` it also defines a vertical barrier, if `-0.9` is close enough to `-1` (the price of 10% of the trade open price cannot be reached in practice). So I do not see any difference between `-1` case and `-0.9`, it should not be "supported", it should be handled properly by backtesting. In this particular case, the close price should be also used:

![2](https://user-images.githubusercontent.com/47309513/70853304-38ded000-1ebd-11ea-95c3-f46d40908069.png)

If "vertical border" is not at the beginning of a candle, the `low` price should be used ('low', to have backtesting pessimistic about the result of this trade)

![3](https://user-images.githubusercontent.com/47309513/70853347-c4f0f780-1ebd-11ea-8db4-fe825f8905e8.png)

(To be continued...)

**Comment by xmatthias** [comment]
to be honest, using -0.9 is just trying to break stuff. 
We can check at startup if all ROI values is -1 or positive - and this problem is solved.

**Comment by polakenfold** [comment]
hi
sorry if comment on closed issue,
i also want to exit after 1day if profit is <0
but doesn't take into account that seting

my startegy:
minimal_roi = {
        "1500": -1,
        "1440": 0.01,
        "80": 0.02,
        "40": 0.03,
        "20": 0.04,
        "0":  0.05
    }

telegram config shows:
Exchange: binance
Stake per trade: unlimited BTC
Minimum ROI: {'1440': 0.01, '80': 0.02, '40': 0.03, '20': 0.04, '0': 0.05}
Stoploss: -0.1
Timeframe: 5m
Strategy: Strategy005b

a0

**Comment by xmatthias** [comment]
Please don't highjack old issues. 
This will notify everyone in this issue.
